packages/components/FocusZone/macos/RCTFocusZone.mm

```diff
@@ -0,0 +1,300 @@
+#import "RCTFocusZone.h"
+
+// from <HIToolbox/Events.h>
+enum {
+	kVK_Tab                       = 0x30,
+	kVK_LeftArrow                 = 0x7B,
+	kVK_RightArrow                = 0x7C,
+	kVK_DownArrow                 = 0x7D,
+	kVK_UpArrow                   = 0x7E
+};
+
+typedef enum {
+	FocusZoneActionNone,
+	FocusZoneActionTab,
+	FocusZoneActionShiftTab,
+	FocusZoneActionRightArrow,
+	FocusZoneActionLeftArrow,
+	FocusZoneActionDownArrow,
+	FocusZoneActionUpArrow,
+} FocusZoneAction;
+
+static const CGFloat FocusZoneBuffer = 3;
+
+@implementation RCTFocusZone
+
+static inline CGFloat GetDistanceBetweenPoints(NSPoint point1, NSPoint point2)
+{
+	NSPoint delta = NSMakePoint(point1.x - point2.x, point1.y - point2.y);
+	return sqrt(delta.x * delta.x + delta.y * delta.y);
+}
+
+static inline NSPoint GetCenterOfRect(NSRect rect)
+{
+	return NSMakePoint(NSMidX(rect), NSMidY(rect));
+}
+
+static inline CGFloat GetDistanceBetweenCentersOfRects(NSRect rect1, NSRect rect2)
+{
+	return GetDistanceBetweenPoints(GetCenterOfRect(rect1), GetCenterOfRect(rect2));
+}
+
+static inline CGFloat GetMinDistanceBetweenRectVerticesAndPoint(NSRect rect, NSPoint point)
+{
+	return fmin(
+		fmin(GetDistanceBetweenPoints(point, NSMakePoint(NSMinX(rect), NSMinY(rect))),
+			GetDistanceBetweenPoints(point, NSMakePoint(NSMaxX(rect), NSMaxY(rect)))),
+		fmin(GetDistanceBetweenPoints(point, NSMakePoint(NSMaxX(rect), NSMinY(rect))),
+			GetDistanceBetweenPoints(point, NSMakePoint(NSMinX(rect), NSMaxY(rect))))
+	);
+}
+
+static NSView *GetFirstResponder(NSWindow *window)
+{
+	NSResponder *responder = [window firstResponder];
+	while (responder != nil && ![responder isKindOfClass:[NSView class]])
+		responder = [responder nextResponder];
+	return [responder isKindOfClass:[NSView class]] ? (NSView *)responder : nil;
+}
+
+static void EnumerateFocusableViews(NSView *root, void (^block)(NSView *, BOOL *))
+{
+	NSMutableArray<NSView *> *queue = [NSMutableArray new] ;
+	[queue addObject:root];
+
+	while ([queue count] > 0)
+	{
+		NSView *view = [queue firstObject];
+		[queue removeObjectAtIndex:0];
+
+		if ([view canBecomeKeyView])
+		{
+			BOOL stop = NO;
+			block(view, &stop);
+			if (stop)
+			{
+				break;
+			}
+		}
+		else
+		{
+			[queue addObjectsFromArray:[view subviews]];
```

Comment:
We should also grab Daniel's CL# 22682075 that deals with nested focusable views in a focusable view?

Reply:
does addobjectsfromarray accept nil? or does it throw exception if [view subviews] is nil?